File: rag-langchain-python/ask_vector_query.py
# Based on example at
# https://python.langchain.com/docs/use_cases/question_answering/quickstart .

import os
import sys
from dotenv import load_dotenv
from langchain import hub
from langchain_openai import AzureChatOpenAI, AzureOpenAIEmbeddings
from langchain.schema import StrOutputParser
from langchain.schema.runnable import RunnablePassthrough
from marklogic import Client
from vector_query_retriever import VectorQueryRetriever


def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)


load_dotenv("../.env")

retriever = VectorQueryRetriever.create(
    Client("http://localhost:8003", digest=("ai-examples-user", "password")),
    embedding_generator=AzureOpenAIEmbeddings(
        azure_deployment=os.environ["AZURE_EMBEDDING_DEPLOYMENT_NAME"]
    )
)

question = sys.argv[1]

fake_langsmith_api_key_to_avoid_warning = "not required for this example"
prompt = hub.pull("rlm/rag-prompt", api_key=fake_langsmith_api_key_to_avoid_warning)

# Note that the Azure OpenAI API key, the Azure OpenAI Endpoint, and the OpenAI API
# Version, are all read from the environment automatically.
llm = AzureChatOpenAI(
    model_name=os.getenv("AZURE_LLM_DEPLOYMENT_NAME"),
    azure_deployment=os.getenv("AZURE_LLM_DEPLOYMENT_NAME"),
    temperature=0,
    max_tokens=None,
    timeout=None,
)

rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | prompt
    | llm
    | StrOutputParser()
)
print(rag_chain.invoke(question))

File: rag-langchain-js/askVectorQuery.js
import { config } from 'dotenv';
import { AzureChatOpenAI } from "@langchain/openai";

import {StringOutputParser} from "@langchain/core/output_parsers";
import {createStuffDocumentsChain} from "langchain/chains/combine_documents";
import * as hub from "langchain/hub";
import {createDatabaseClient} from "marklogic";
import {VectorQueryRetriever} from "./vectorQueryRetriever.js";

config({path: "../.env"});
const marklogicClient = createDatabaseClient({
  host:     'localhost',
  port:     '8003',
  user:     'ai-examples-user',
  password: 'password',
  authType: 'DIGEST'
});
const vectorQueryRetriever = new VectorQueryRetriever({marklogicClient});

// We need to specify the environment variables since the Python and Java examples use different default variables.
const llm = new AzureChatOpenAI({
  azureOpenAIApiVersion: process.env.OPENAI_API_VERSION,
  azureOpenAIApiDeploymentName: process.env.AZURE_LLM_DEPLOYMENT_NAME
});
const prompt = await hub.pull("rlm/rag-prompt");

let question = "What disturbances has Jane Doe caused?";
if (process.argv.length > 2) {
  question = process.argv[2];
}
console.log(`Question: ${question}`);

const ragChain = await createStuffDocumentsChain({
  llm,
  prompt,
  outputParser: new StringOutputParser()
});
const chainResponse = await ragChain.invoke({
  question: question,
  context: await vectorQueryRetriever.invoke(question),
});
console.log("\n\nChatbot VectorQuery chainResponse: \n" + JSON.stringify(chainResponse));
